### ISSUE TITLE
Naive thunk closures cause space leaks

### ISSUE THREAD

**Issue by roberth** [NixOS/nix#8285]
**Is your feature request related to a problem? Please describe.**

The interpreter uses a naive (but perhaps time-efficient) representation of closures, which is to represent the entire scope using a singly linked "immutable" list of `Env`s, each representing a "level" of scope, starting with the innermost, so that the outermost scopes can be shared. (Only benign, idempotent mutation happens, from thunk to proper value)
This works, but it means that a closure needs to hold a reference to the entire scope, even though only specific variables in that scope will ever be accessed. This leads to unnecessary space leaks: whenever an expression returns a thunk, all other variables in that scope are retained until all such thunks have been evaluated successfully.
We already have individually collectable thunks. `Env` just happens to point to all of them, and our thunk point to the `Env`s.

**Describe the solution you'd like**

An alternative implementation could:

1. Have two `ExprVar` lookup strategies.
   - the current strategy for large sets of free variables, which would be "too big to copy". Example: `rec { ..... }` scopes.
   - a new copying strategy which performs a lookup in a small flattened closure of up to an experimentally determined number of variables
2. During parsing, maintain the set of free variables for each node.
3. During scope checking (bindVars), decide which closure type to use, save that decision, and save the appropriate index in `ExprVar`. The set of free variables may be cleared to release any extra memory that was used.

**Describe alternatives you've considered**

Status quo is simple and possibly time efficient for small expressions.
An oversized heap does incur recurring time losses though, due to slower GC!

**Additional context**

This would be an experiment that may fail to produce a positive result.

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by roberth** [comment]
The naive closure representation is time-efficient.
Ideally we could compact them during gc
- https://github.com/NixOS/nix/issues/8626#issuecomment-3245493665

**Comment by wingo** [comment]
Just for what it is worth: I don't think you want to compact during GC.  GC-time is very weird, you can observe references in both moved and pre-moved states, the "compaction" actually might require more memory than before, other threads might be moving objects that the closure refers to, etc.  If it were me I would produce display closures (closures that just reference needed values) at some other time: either eagerly, or when a lazy value is evaluated, or during a separate compile phase.  Just FWIW :)

**Comment by roberth** [comment]
Yeah, interleaving that somehow seems like asking for trouble, at least with the amount of control I think we have with boehm, but perhaps a good time is either the alternatives you suggest, or between stopping the world and the actual GC work. *Maybe* integrating into marking is possible? idk. Also this compaction is kind of like moving, so if, say, MMTk is very very flexible, I can sort of imagine it to work. My imagination does not come with a warranty though ;)
